src/hotspot/share/utilities/events.hpp

```diff
@@ -414,16 +428,48 @@ inline void EventLogBase<ExtendedStringLogMessage>::print(outputStream* out, Ext
   out->cr();
 }
 
+typedef void (*EventLogFunction)(Thread* thread, const char* format, ...);
+
+class EventMarkBase : public StackObj {
+  EventLogFunction _log_function;
+  StringLogMessage _buffer;
+
+ protected:
+  void log_start(const char* format, va_list argp) ATTRIBUTE_PRINTF(2, 0);
+  void log_end();
+
+ public:
+  EventMarkBase(EventLogFunction log_function);
+};
```

Comment:
The `EventMarkBase` constructor could be protected as well, and the class `NONCOPYABLE`.

Reply:
Will update. Thanks for reviewing.